### PR TITLE
[HttpFoundation] Precalculate session expiry timestamp

### PR DESCRIPTION
**Pull request by azjezz**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no  
| Deprecations? | yes
| Tests pass?   | yes  
| Fixed tickets | #13955, #21423, #14625
| License       | MIT

Continued work from the original PR #21423 / #21857

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
The migration path doesn't work IMHO.
Instead, I propose to keep the name of the column and of the property as is.
Then, the column should be indexed, and we should use it to store absolute expiries.
To have smooth compat with current sessions, to decide when a value in the column is a lifetime or an expiry, we would compare values to `315576000` (10 years): when the number is higher, it's an absolute date, when it's ower, it's a relative date.
GC would be done in two DELETE queries, for >315576000 and <=315576000.
In v6, we would simply drop the <=315576000 one, because in two years we'll be able to consider that all relevant sessions have been migrated to absolute dates.

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php [hunk 1]
```diff
@@ -403,13 +398,14 @@ public function close()
 
             // delete the session records that have expired
             if ('mysql' === $this->driver) {
-                $sql = "DELETE FROM $this->table WHERE $this->lifetimeCol + $this->timeCol < :time";
+                $sql = "DELETE FROM $this->table WHERE ($this->lifetimeCol < :time AND $this->lifetimeCol > :min) OR ($this->lifetimeCol + $this->timeCol < :time)";
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: using `OR` means that the DB engine won't use the index AFAIK
that's why I suggested using two queries

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php [hunk 2]
```diff
@@ -403,13 +398,14 @@ public function close()
...
             } else {
-                $sql = "DELETE FROM $this->table WHERE $this->lifetimeCol < :time - $this->timeCol";
+                $sql = "DELETE FROM $this->table WHERE ($this->lifetimeCol < :time AND $this->lifetimeCol > :min) OR ($this->lifetimeCol < :time - $this->timeCol)";
             }
 
             $stmt = $this->pdo->prepare($sql);
             $stmt->bindValue(':time', time(), \PDO::PARAM_INT);
+            $stmt->bindValue(':min', 315576000, \PDO::PARAM_INT);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: let's use a private const for clarifty

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php [hunk 3]
```diff
@@ -837,32 +851,32 @@ private function getMergeStatement(string $sessionId, string $data, int $maxlife
...
             case 'pgsql' === $this->driver && version_compare($this->pdo->getAttribute(\PDO::ATTR_SERVER_VERSION), '9.5', '>='):
-                $mergeSql = "INSERT INTO $this->table ($this->idCol, $this->dataCol, $this->lifetimeCol, $this->timeCol) VALUES (:id, :data, :lifetime, :time) ".
+                $mergeSql = "INSERT INTO $this->table ($this->idCol, $this->dataCol, $this->lifetimeCol, $this->timeCol) VALUES (:id, :data, :expiry, :time) ".
                     "ON CONFLICT ($this->idCol) DO UPDATE SET ($this->dataCol, $this->lifetimeCol, $this->timeCol) = (EXCLUDED.$this->dataCol, EXCLUDED.$this->lifetimeCol, EXCLUDED.$this->timeCol)";
                 break;
             default:
                 // MERGE is not supported with LOBs: https://oracle.com/technetwork/articles/fuecks-lobs-095315.html
                 return null;
-        }
+            }
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: bad indentation

**Review by nicolas-grekas — APPROVED** [review]
LGTM thanks, just one minor comment

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php [hunk 4]
```diff
@@ -237,6 +239,7 @@ public function createTable()
 
         try {
             $this->pdo->exec($sql);
+            $this->pdo->exec("CREATE INDEX EXPIRY ON $this->table ($this->lifetimeCol)");
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: we could add this query (with default table+col) in UPGRADE4.4&5.0 files to hint ppl they should add the index

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php [hunk 5]
```diff
@@ -401,15 +404,21 @@ public function close()
...
             } else {
-                $sql = "DELETE FROM $this->table WHERE $this->lifetimeCol < :time - $this->timeCol";
+                $legacySql = "DELETE FROM $this->table WHERE $this->lifetimeCol <= :min AND $this->lifetimeCol < :time - $this->timeCol";
             }
 
             $stmt = $this->pdo->prepare($sql);
             $stmt->bindValue(':time', time(), \PDO::PARAM_INT);
+            $stmt->bindValue(':min', self::MAX_LIFETIME, \PDO::PARAM_INT);
+            $stmt->execute();
+            $stmt = $this->pdo->prepare($legacySql);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: with this comment before: `// to be removed in 6.0`

File: UPGRADE-4.4.md [hunk 6]
```diff
@@ -109,6 +109,7 @@ HttpFoundation
 
  * `ApacheRequest` is deprecated, use `Request` class instead.
  * Passing a third argument to `HeaderBag::get()` is deprecated since Symfony 4.4, use method `all()` instead
+ * `PdoSessionHandler` now precalculates the expiry timestamp in the lifetime column, make sure to run `CREATE INDEX EXPIRY ON sessions (sess_lifetime)` to update your database.
```
Comment (Tobion) [inline UPGRADE-4.4.md]: You should add why it's recommended to add the index: to speed up gc of expired sessions

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php [hunk 7]
```diff
@@ -616,7 +626,12 @@ protected function doRead($sessionId)
             $sessionRows = $selectStmt->fetchAll(\PDO::FETCH_NUM);
 
             if ($sessionRows) {
-                if ($sessionRows[0][1] + $sessionRows[0][2] < time()) {
+                $expiry = (int) $sessionRows[0][1];
+                if ($expiry <= self::MAX_LIFETIME) {
+                    $expiry += $sessionRows[0][2];
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php]: this also means, we do not need to fetch the time column data in the future at all. please add a comment on the select sql so we remember

**Review by Tobion — CHANGES_REQUESTED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @azjezz.

**Comment by tseho** [comment]
Hello @nicolas-grekas, I think this is a BC break. On mysql, the lifetime column has been for the last few years a MEDIUMINT column.

The new version with `time() + $maxlifetime` instead of `$maxlifetime` is higher than the MEDIUMINT mysql limit.

When upgrading from symfony 4.3 to 4.4 with an existing database, we have the error `SQLSTATE[22003]: Numeric value out of range: 1264 Out of range value for column 'sess_lifetime' at row 1`

I saw the creation of the session table has been fixed 5 days ago in https://github.com/symfony/symfony/pull/34410/files but that does not fix existing databases.

Is it normal we need to create manually a sql migration for a breaking database schema change, which is not really mentioned in UPGRADE-4.4.md ?

**Comment by nicolas-grekas** [comment]
That's an interesting feedback @tseho but please create an issue: nobody is tracking comments on closed PRs.

**Comment by tseho** [comment]
Thanks for your response and sorry about that, I just created a new issue : https://github.com/symfony/symfony/issues/34491

**Pull request merged**